reps/2022-04-07-serialization.md

```diff
@@ -0,0 +1,162 @@
+#### Implement Serializer
+Then, he needs to implement this custom serializer, implements the interface shown below:
+Let's take Python as an example. Other languages will be similar.
+```python
+class MySerializer(RaySerializer):
+    def serialize(self, object: MyClass) -> RaySerializationResult:
+        pass
+    def deserialize(self, in_band_data: bytes,
+                    out_of_band_data: List[bytes]) -> MyClass:
+        pass
+
+```
+`serialize`method should return a `SerializedResult`, which contains 2 fields: in-band buffer and out-of-band buffers.
+`inBandBuffer` is nothing else than a byte array. Users can use this field to achieve normal in-band serialization.
+```python
+class RaySerializationResult:
+    def __init__(self):
+        self.in_band_buffer: bytes = None
+        self.out_of_band_buffers: List[RayOutOfBandBUffer] = None
+```
+`OutOfBandBuffer` is for advanced users. They can use this to do out-of-band optimization. Ray will call `OutOfBandBuffer#write` only once when constructing the final buffer sent to the network.
+```python
+class MyOutOfBandBuffer(RayOutOfBandBUffer):
+    def get_size(self) -> int:
+        pass
+    # dest is a byte array with size == get_size()
+    def write_to(self, dest: bytes):
+        pass
+```
+### Final Buffer Protocol
+```
++---------------+----------------+-------------------------------------
+|  ClassIDHash  |  in-band data  |   … oob-buf0 … oob-buf1 … oob-buf2 …
++---------------+----------------+-------------------------------------
+```
+A msgpack buffer which includes 3 parts:
+The 1st part is the class ID which is used to find the corresponding serializer.
+The 2nd part is the serialized data of the real object. 
+The last part is a list of out-of-band buffers in memory.
+#### Examples
+##### Custom Object with OOB Fields
+```java
+ClassWithOOB {
+    int a; // in-band serialized
+    byte[] d; // out-of-band serialized
+}
+```
+Here if a user imlements a serializer for this class, he will put `int a` to in-band buffer, and `byte[] d` to the out-of-band buffer.
+The buffer layout will be:
+
+| Class ID hash | hash("ClassWithOOB") |
+| --- | --- |
+| In-band data | a |
+| Out of band data list | [d] |
+
+When we get this buffer in another language, we'll firstly check whether a serializer for "ClassWithOOB" is registered. If not, an serialization exception will be thrown. Else, we put in-band and out-of-band data to the `deseriaze` method of that serializer, get the result and continue processing.
+### Work Items
+#### Refactor Serialization Code Path
+Unify all existing object serialization to this API.
+
+1. Implement internal object serializer, such as RayError, ActorHandle, etc.
+1. By default, if no serializer is registered, use current serializer as the fallback. For example, pickle5 in Python, FST in Java. **In this case, cross-language serialization is disabled.**
```

Comment:
Can we avoid this regression? For example, you can register Msgpack as the serializer for the primitive types.